rfcs/0161-deploymentid.md

```diff
@@ -0,0 +1,112 @@
+for an unbounded period of time.
+
+With intelligent worker cycling, we can ensure that workers with outdated
+configuration are swiftly removed from worker pools.
+
+# Details
+
+The responsibility for keeping workers up-to-date is split between Worker
+Manager (for ensuring that provider launch configuration of workers is valid)
+and the workers, responsible for refreshing `workerConfig` on reregisterWorker
+calls.
+
+This RFC involves changes to both Worker Manager and Worker-Managed-spawned
+workers (via Worker Runner):
+
+* Worker Manager will keep a record of the launch parameters it used for all
+  workers that it spawns
+* Worker Manager will track changes to Worker Pool definitions that affect
+  launch configurations of workers
+* When Worker Manager receives a `workermanager.registerWorker` or
+  `workermanager.reregisterWorker` call from a worker whose launch
+  configuration is determined to be from a previous version of the current
+  worker pool definition, it will assess whether its launch configuration is
+  still valid against the latest worker pool definition, and if not, return a
+  HTTP 410 status code with a message body explaining that the worker's launch
+  configuration is no longer conformant.
+* Worker Manager will include (the most recent) `workerConfig` in
+  `workermanager.reregisterWorker` responses for workers with up-to-date launch
+  configurations (`workermanager.registerWorker` already includes `workerConfig`)
+* Workers (via Worker Runner) will refresh their configs from successful
+  `workermanager.reregisterWorker` calls, or gracefully shutdown when receiving
+  a HTTP 410 status code from either `workermanager.registerWorker` or
+  `workermanager.reregisterWorker`
+  
+
+## Worker Manager changes
+
+Depending on the provider, Worker Pools contain different data. Regardless, the
+logic to decide if a worker pool launch configuration has changed is the same:
+
+* If Worker Manager can determine that the parameters used to launch an
+  existing running worker (instance type, disk sizes, region, ...) from a
+  previous version of the worker pool definition are still a valid combination
+  of parameters against the latest worker pool configuration, responses to
+  `workermanager.registerWorker` or `workermanager.reregisterWorker` will be
+  treated as before, and the worker will not be earmarked for decommission.
+* Otherwise, if Worker Manager is either unable to determine if the launch
+  parameters of a given worker are still valid, or is able to determine that
+  they are no longer consisten with the latest worker pool definition, it should:
+
+    * Respond to `workermanager.registerWorker` and `workermanager.reregisterWorker`
+      calls with an HTTP 410 response status code
+    * Terminate the worker (in the case of non-static workers) as soon as
+      possible, but _no earlier_ than 15 minutes after the worker's current
+      credentials have expired
```

Comment:
I'm not sure I understand the motivation for the restriction on timing here. I also wonder if the quarantine provision below should always happen, independent of whether the worker is static or not.

Reply:
It is mostly a hard-stop for the worker, to give it a period to gracefully terminate. The 15 minutes acts as a rather generous offset for any click-drift between worker and service, assuming that if the worker is still running 15 minutes after its credentials became invalid, it will no longer be able to successfully interact with guarded API endpoints of taskcluster. Perhaps though, the existing machinery of Worker Manager to terminate invalid workers would automatically  be engaged once the worker is removed from the active pool, so perhaps no additional changes are needed. @imbstack can you confirm?